### ISSUE TITLE
HAMT interface

### ISSUE THREAD

**Issue by ruslansennov** [javaslang/javaslang#591]
Nothing will extend it in the future and we can convert it to abstract class to hide inner classes.
What do you think about it?

**Comment by danieldietrich** [comment]
I don't like abstract classes as first class API. Let's do the following instead which, according to javadoc, also works to hide classes from public API: We move the inner classes to the outside within the same file.

```
public interface HashArrayMappedTrie<K, V> extends java.lang.Iterable<Tuple2<K, V>> {
    ...
}

abstract class AbstractNode<K, V> implements HashArrayMappedTrie<K, V> {
    ...
}

class ArrayNode ... { ... }

...
```

**Comment by danieldietrich** [comment]
I will do this with #592, which will hide all classes all over our library, which are not intended to be public api.

**Comment by danieldietrich** [comment]
Fixed with #610
